common/engine_conf/servicedependency_helper.cc

```diff
@@ -196,4 +196,142 @@ void servicedependency_helper::_init() {
+          dep.dependent_hosts().data(), dep.dependent_hostgroups().data(),
+          dep.dependent_service_description().data(),
+          dep.dependent_servicegroups().data(), s, dependent_services);
+
+      // Browse all depended and dependent services.
+      for (auto& p1 : depended_services)
+        for (auto& p2 : dependent_services) {
+          // Create service dependency instance.
+          for (int32_t i = 1; i <= 2; i++) {
+            if (dep.dependency_type() == DependencyKind::unknown ||
+                static_cast<int32_t>(dep.dependency_type()) == i) {
+              auto sdep = std::make_unique<Servicedependency>();
+              sdep->CopyFrom(dep);
+              sdep->clear_hostgroups();
+              sdep->clear_hosts();
+              sdep->mutable_hosts()->add_data(p1.first);
+              sdep->clear_servicegroups();
+              sdep->clear_service_description();
+              sdep->mutable_service_description()->add_data(p1.second);
+              sdep->clear_dependent_hostgroups();
+              sdep->clear_dependent_hosts();
+              sdep->mutable_dependent_hosts()->add_data(p2.first);
+              sdep->clear_dependent_servicegroups();
+              sdep->clear_dependent_service_description();
+              sdep->mutable_dependent_service_description()->add_data(
+                  p2.second);
+              if (i == 2) {
+                sdep->set_dependency_type(DependencyKind::execution_dependency);
+                sdep->set_notification_failure_options(0);
+              } else {
+                sdep->set_dependency_type(
+                    DependencyKind::notification_dependency);
+                sdep->set_execution_failure_options(0);
+              }
+              expanded.push_back(std::move(sdep));
+            }
+          }
+        }
+    }
+  }
+
+  // Set expanded service dependencies in configuration state.
+  s.clear_servicedependencies();
+  for (auto& e : expanded)
+    s.mutable_servicedependencies()->AddAllocated(e.release());
+}
+
+/**
+ * @brief Expand services.
+ *
+ * @param hst Hosts.
+ * @param hg Host groups.
+ * @param svc Service descriptions.
+ * @param sg Service groups.
+ * @param s Configuration state.
+ * @param expanded Expanded services.
+ */
+void servicedependency_helper::_expand_services(
+    const ::google::protobuf::RepeatedPtrField<std::string>& hst,
+    const ::google::protobuf::RepeatedPtrField<std::string>& hg,
+    const ::google::protobuf::RepeatedPtrField<std::string>& svc,
+    const ::google::protobuf::RepeatedPtrField<std::string>& sg,
+    configuration::State& s,
+    absl::flat_hash_set<std::pair<std::string, std::string>>& expanded) {
+  // Expanded hosts.
+  absl::flat_hash_set<std::string> all_hosts;
+
+  // Base hosts.
+  all_hosts.insert(hst.begin(), hst.end());
+
+  // Host groups.
+  for (auto& hgn : hg) {
+    // Find host group
+    auto found = std::find_if(
+        s.hostgroups().begin(), s.hostgroups().end(),
```

Comment:
I see many std::find_if, it wouldn't be better to store some protobuf data in map instead of repeated. I'm a little afraid by cpu load on big configs